pkg/api/vmtemplate/store.go

```diff
@@ -0,0 +1,121 @@
+package vmtemplate
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/vm/pkg/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/rancher/apiserver/pkg/types"
+	ctlvmv1alpha1 "github.com/rancher/vm/pkg/generated/controllers/vm.cattle.io/v1alpha1"
+	"github.com/rancher/vm/pkg/ref"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+)
+
+const (
+	templateLabel = "template.vm.cattle.io/templateID"
+)
+
+type templateVersionStore struct {
+	types.Store
+	templateCache        ctlvmv1alpha1.TemplateCache
+	templateVersions     ctlvmv1alpha1.TemplateVersionClient
+	templateVersionCache ctlvmv1alpha1.TemplateVersionCache
+	mu                   sync.Mutex
+}
+
+func (s *templateVersionStore) Create(request *types.APIRequest, schema *types.APISchema, data types.APIObject) (types.APIObject, error) {
+	ns := data.Data().String("metadata", "namespace")
+	if ns == "" {
+		data.Data().SetNested(config.Namespace, "metadata", "namespace")
+	}
+
+	templateID := data.Data().String("spec", "templateId")
+	if templateID == "" {
+		return types.APIObject{}, errors.New("TemplateId is empty")
```

Comment:
Can we return a proper apierror, otherwise it returns 500 when input is invalid. Check other places as well.

Reply:
updated.